### PR TITLE
Using of try/catch to fix normalized identifier exception

### LINKED ISSUE #6535: Can not get the normalized identifier for App\Entity\BettingEventCategory since it is in state 2.
#### Sonata packages

```
$ composer show --latest 'sonata-project/*'
sonata-project/admin-bundle              dev-master 890ecee dev-master 8773923 The missing Symfony Admin Generator
sonata-project/block-bundle              4.4.0              4.4.0              Symfony SonataBlockBundle
sonata-project/cache                     2.0.1              2.0.1              Cache library
sonata-project/doctrine-extensions       1.10.1             1.10.1             Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle dev-master a4391d3 dev-master a4391d3 Integrate Doctrine ORM into the Sona...
sonata-project/exporter                  2.4.1              2.4.1              Lightweight Exporter library
sonata-project/form-extensions           1.6.0              1.6.0              Symfony form extensions
sonata-project/twig-extensions           1.4.1              1.4.1              Sonata twig extensions
```

#### Symfony packages

```
$ composer show --latest 'symfony/*'
symfony/apache-pack                      v1.0.1             v1.0.1             A pack for Apache support in Symfony
symfony/asset                            v5.1.7             v5.1.7             Symfony Asset Component
symfony/browser-kit                      v5.1.7             v5.1.7             Symfony BrowserKit Component
symfony/cache                            v5.1.7             v5.1.7             Symfony Cache component with PSR-6, ...
symfony/cache-contracts                  v2.2.0             v2.2.0             Generic abstractions related to caching
symfony/config                           v5.1.7             v5.1.7             Symfony Config Component
symfony/console                          v5.1.7             v5.1.7             Symfony Console Component
symfony/css-selector                     v5.1.7             v5.1.7             Symfony CssSelector Component
symfony/debug-bundle                     v5.1.7             v5.1.7             Symfony DebugBundle
symfony/dependency-injection             v5.1.7             v5.1.7             Symfony DependencyInjection Component
symfony/deprecation-contracts            v2.2.0             v2.2.0             A generic function and convention to...
symfony/doctrine-bridge                  v5.1.7             v5.1.7             Symfony Doctrine Bridge
symfony/dom-crawler                      v5.1.7             v5.1.7             Symfony DomCrawler Component
symfony/dotenv                           v5.1.7             v5.1.7             Registers environment variables from...
symfony/error-handler                    v5.1.7             v5.1.7             Symfony ErrorHandler Component
symfony/event-dispatcher                 v5.1.7             v5.1.7             Symfony EventDispatcher Component
symfony/event-dispatcher-contracts       v2.2.0             v2.2.0             Generic abstractions related to disp...
symfony/expression-language              v5.1.7             v5.1.7             Symfony ExpressionLanguage Component
symfony/filesystem                       v5.1.7             v5.1.7             Symfony Filesystem Component
symfony/finder                           v5.1.7             v5.1.7             Symfony Finder Component
symfony/flex                             v1.9.10            v1.9.10            Composer plugin for Symfony
symfony/form                             v5.1.7             v5.1.7             Symfony Form Component
symfony/framework-bundle                 v5.1.7             v5.1.7             Symfony FrameworkBundle
symfony/http-client                      v5.1.7             v5.1.7             Symfony HttpClient component
symfony/http-client-contracts            v2.3.1             v2.3.1             Generic abstractions related to HTTP...
symfony/http-foundation                  v5.1.7             v5.1.7             Symfony HttpFoundation Component
symfony/http-kernel                      v5.1.7             v5.1.7             Symfony HttpKernel Component
symfony/intl                             v5.1.7             v5.1.7             A PHP replacement layer for the C in...
symfony/mailer                           v5.1.7             v5.1.7             Symfony Mailer Component
symfony/maker-bundle                     v1.22.0            v1.22.0            Symfony Maker helps you create empty...
symfony/mime                             v5.1.7             v5.1.7             A library to manipulate MIME messages
symfony/monolog-bridge                   v5.1.7             v5.1.7             Symfony Monolog Bridge
symfony/monolog-bundle                   v3.6.0             v3.6.0             Symfony MonologBundle
symfony/notifier                         v5.1.7             v5.1.7             A library to notify messages
symfony/options-resolver                 v5.1.7             v5.1.7             Symfony OptionsResolver Component
symfony/phpunit-bridge                   v5.1.7             v5.1.7             Symfony PHPUnit Bridge
symfony/polyfill-intl-grapheme           v1.18.1            v1.19.0            Symfony polyfill for intl's grapheme...
symfony/polyfill-intl-icu                v1.18.1            v1.19.0            Symfony polyfill for intl's ICU-rela...
symfony/polyfill-intl-idn                v1.18.1            v1.19.0            Symfony polyfill for intl's idn_to_a...
symfony/polyfill-intl-normalizer         v1.18.1            v1.19.0            Symfony polyfill for intl's Normaliz...
symfony/polyfill-mbstring                v1.18.1            v1.19.0            Symfony polyfill for the Mbstring ex...
symfony/polyfill-php73                   v1.18.1            v1.19.0            Symfony polyfill backporting some PH...
symfony/polyfill-php80                   v1.18.1            v1.19.0            Symfony polyfill backporting some PH...
symfony/process                          v5.1.7             v5.1.7             Symfony Process Component
symfony/property-access                  v5.1.7             v5.1.7             Symfony PropertyAccess Component
symfony/property-info                    v5.1.7             v5.1.7             Symfony Property Info Component
symfony/routing                          v5.1.7             v5.1.7             Symfony Routing Component
symfony/security-acl                     v3.1.0             v3.1.0             Symfony Security Component - ACL (Ac...
symfony/security-bundle                  v5.1.7             v5.1.7             Symfony SecurityBundle
symfony/security-core                    v5.1.7             v5.1.7             Symfony Security Component - Core Li...
symfony/security-csrf                    v5.1.7             v5.1.7             Symfony Security Component - CSRF Li...
symfony/security-guard                   v5.1.7             v5.1.7             Symfony Security Component - Guard
symfony/security-http                    v5.1.7             v5.1.7             Symfony Security Component - HTTP In...
symfony/serializer                       v5.1.7             v5.1.7             Symfony Serializer Component
symfony/service-contracts                v2.2.0             v2.2.0             Generic abstractions related to writ...
symfony/stopwatch                        v5.1.7             v5.1.7             Symfony Stopwatch Component
symfony/string                           v5.1.7             v5.1.7             Symfony String component
symfony/translation                      v5.1.7             v5.1.7             Symfony Translation Component
symfony/translation-contracts            v2.3.0             v2.3.0             Generic abstractions related to tran...
symfony/twig-bridge                      v5.1.7             v5.1.7             Symfony Twig Bridge
symfony/twig-bundle                      v5.1.7             v5.1.7             Symfony TwigBundle
symfony/validator                        v5.1.7             v5.1.7             Symfony Validator Component
symfony/var-dumper                       v5.1.7             v5.1.7             Symfony mechanism for exploring and ...
symfony/var-exporter                     v5.1.7             v5.1.7             A blend of var_export() + serialize(...
symfony/web-link                         v5.1.7             v5.1.7             Symfony WebLink Component
symfony/web-profiler-bundle              v5.1.7             v5.1.7             Symfony WebProfilerBundle
symfony/webpack-encore-bundle            v1.7.3             v1.7.3             Integration with your Symfony app & ...
symfony/yaml                             v5.1.7             v5.1.7             Symfony Yaml Component
```

#### PHP version

```
$ php -v
PHP 7.3.12 (cli) (built: Nov 19 2019 13:58:02) ( ZTS MSVC15 (Visual C++ 2017) x64 )
```

## Subject

Can not get the normalized identifier for App\Entity\BettingEventCategory since it is in state 2.

## Steps to reproduce

BettingEventCategory class

```php
<?php

namespace App\Entity;

use App\Repository\BettingEventCategoryRepository;
use Doctrine\ORM\Mapping as ORM;

/**
 * @ORM\Entity(repositoryClass=BettingEventCategoryRepository::class)
 */
class BettingEventCategory
{
    /**
     * @ORM\Id
     * @ORM\GeneratedValue
     * @ORM\Column(type="integer")
     */
    private $id;

    /**
     * @ORM\Column(type="string", length=255)
     */
    private $categoryName;

    /**
     * @ORM\Column(type="boolean")
     */
    private $enabled;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getCategoryName(): ?string
    {
        return $this->categoryName;
    }

    public function setCategoryName(string $categoryName): self
    {
        $this->categoryName = $categoryName;

        return $this;
    }

    public function getEnabled(): ?bool
    {
        return $this->enabled;
    }

    public function setEnabled(bool $enabled): self
    {
        $this->enabled = $enabled;

        return $this;
    }
}
```

BettingEventCategoryAdmin  class
```php
<?php

namespace App\Admin;

use Sonata\AdminBundle\Admin\AbstractAdmin;
use Sonata\AdminBundle\Datagrid\DatagridMapper;
use Sonata\AdminBundle\Datagrid\ListMapper;
use Sonata\AdminBundle\Form\FormMapper;
use Sonata\Form\Type\BooleanType;
use Symfony\Component\Form\Extension\Core\Type\TextType;

class BettingEventCategoryAdmin extends AbstractAdmin
{
    protected function configureFormFields(FormMapper $formMapper): void
    {
        $formMapper
            ->add('categoryName', TextType::class)
            ->add('enabled', BooleanType::class, array(
                'label' => 'Enabled',
                'transform' => true,
                'required' => true,
            ));
    }

    protected function configureDataGridFilters(DatagridMapper $datagridMapper): void
    {
        $datagridMapper
            ->add('id')
            ->add('categoryName')
            ->add('enabled');
    }

    protected function configureListFields(ListMapper $listMapper): void
    {
        $listMapper
            ->addIdentifier('id')
            ->addIdentifier('categoryName')
            ->addIdentifier('enabled');
    }
}
```

My service section
```yaml
admin.betting_event_category:
    class: App\Admin\BettingEventCategoryAdmin
    arguments: [~, App\Entity\BettingEventCategory, ~]
    tags:
      - { name: sonata.admin, manager_type: orm, label: "Betting Event Category" }
```
Trying to create a new entity
imagen

## Expected results
The new entity should be created

## Actual results

Error received
imagen

```
InvalidArgumentException:
Can not get the normalized identifier for App\Entity\BettingEventCategory since it is in state 2.

  at C:\xampp\htdocs\vendor\sonata-project\doctrine-orm-admin-bundle\src\Model\ModelManager.php:396
  at Sonata\DoctrineORMAdminBundle\Model\ModelManager->getNormalizedIdentifier(object(BettingEventCategory))
     (C:\xampp\htdocs\vendor\sonata-project\admin-bundle\src\Admin\AbstractAdmin.php:2001)
  at Sonata\AdminBundle\Admin\AbstractAdmin->getNormalizedIdentifier(object(BettingEventCategory))
     (C:\xampp\htdocs\vendor\sonata-project\admin-bundle\src\Admin\AbstractAdmin.php:2006)
  at Sonata\AdminBundle\Admin\AbstractAdmin->id(object(BettingEventCategory))
     (C:\xampp\htdocs\vendor\sonata-project\admin-bundle\src\Admin\AbstractAdmin.php:1984)
  at Sonata\AdminBundle\Admin\AbstractAdmin->isGranted('CREATE', object(BettingEventCategory))
     (C:\xampp\htdocs\vendor\sonata-project\admin-bundle\src\Admin\AbstractAdmin.php:2195)
  at Sonata\AdminBundle\Admin\AbstractAdmin->checkAccess('create', object(BettingEventCategory))
     (C:\xampp\htdocs\vendor\sonata-project\admin-bundle\src\Controller\CRUDController.php:549)
  at Sonata\AdminBundle\Controller\CRUDController->createAction(object(Request))
     (C:\xampp\htdocs\vendor\symfony\http-kernel\HttpKernel.php:157)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (C:\xampp\htdocs\vendor\symfony\http-kernel\HttpKernel.php:79)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (C:\xampp\htdocs\vendor\symfony\http-kernel\Kernel.php:196)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (C:\xampp\htdocs\index.php:28) 
```

### PR DESCRIPTION
**Pull request by ggabrovski**
## Subject
I am targeting this branch, because BC.

Using of `try/catch` to fix SonataDoctrineORM exception coming from `getNormalizedIdentifier`

Closes #6535

### Fixed
Fix for can not get the normalized identifier exception.

### CODE REVIEW

File: src/Admin/AbstractAdmin.php [hunk 1]
```diff
@@ -2005,16 +2005,28 @@ public function isGranted($name, ?object $object = null): bool
 
     public function getUrlSafeIdentifier(object $model): string
     {
-        return $this->getModelManager()->getUrlSafeIdentifier($model);
+        try {
+            return $this->getModelManager()->getUrlSafeIdentifier($model);
+        } catch (\InvalidArgumentException $e) {
+            return '';
```
Comment (core23) [inline src/Admin/AbstractAdmin.php:2011]: This looks wrong. If there is a problem to get the identifier, you should see the error

**Review by core23 — CHANGES_REQUESTED** [review]

**Comment by ggabrovski** [comment]
I agree with you, but if we apply the logic described here #6535, just to return null and to make changes in SonataDoctrineORM we will not see the error either. Can you give me some guidelines and I will try to make new pull request. Thank you.

File: src/Admin/AbstractAdmin.php [hunk 2]
```diff
@@ -2005,16 +2005,28 @@ public function isGranted($name, ?object $object = null): bool
...
-        return $this->getModelManager()->getNormalizedIdentifier($model);
+        try {
+            return $this->getModelManager()->getNormalizedIdentifier($model);
+        } catch (\InvalidArgumentException $e) {
+            return '';
+        }
     }
 
-    public function id(object $model): string
+    public function id(?object $model): string
```
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:2024]: This shouldn't be called with `null`.

What's your use case ? There was none in the issue #6535

File: src/Admin/AdminInterface.php [hunk 3]
```diff
@@ -188,7 +188,7 @@ public function getNormalizedIdentifier(object $model): string;
      *
      * @phpstan-param T $model
      */
-    public function id(object $model): string;
+    public function id(?object $model): string;
```
Comment (VincentLanglet) [inline src/Admin/AdminInterface.php:191]: This shouldn't be called with null.

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

**Comment by VincentLanglet** [comment]
TLDR: I recommend to do 3 PR:
- One on SonataAdmin to allow `null` as return type for getUrlSafeIdentifier, getNormalizedIdentifier, and id.
- One on SonataDoctrineORM@3.x to remove the deprecation in getNormalizedIdentifier.
- One on SonataDoctrineORM@master to remove the exception and return null instead.

> I agree with you, but if we apply the logic described here #6535, just to return null and to make changes in SonataDoctrineORM we will not see the error either. Can you give me some guidelines and I will try to make new pull request. Thank you.

Here: https://github.com/sonata-project/SonataAdminBundle/issues/6535#issuecomment-719741227
I explained that the error is coming from the `isGranted` method.

One solution could be to try catch in the `isGranted` method then.
```
$objectRef = $object ? sprintf('/%s#%s', spl_object_hash($object), $this->id($object)) : '';
```
to
```
try {
    $objectRef = $object ? sprintf('/%s#%s', spl_object_hash($object), $this->id($object)) : '';
} catch (..)
    $objectRef = '';
}
```

But I also explained there was something wrong to me with the exception thrown.

I don't expect to catch something when there is no `@throws` on a method, and the interface is 
```
   /**
     * Shorthand method for templating.
     *
     * @phpstan-param T $model
     */
    public function id(object $model): string;
```
I also generally don't expect to catch LogicException.

Either we should add `@throws RuntimeException When the model has no identifier`, either we should allow `null`.

And to me the call `$this->id($foo)` is similar to `$foo->getId()` which returns `null` when the object is new.
So I consider `null` as a valid return value.

That's why I said It seems right to remove the exception in the persistence bundle and allow `null` as return value for 
- getUrlSafeIdentifier
- getNormalizedIdentifier
- id

Then a PR in SonataDoctrineORM 3.x to remove the deprecation
https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/src/Model/ModelManager.php#L456-L469

And a PR in SonataDoctrineORM master to remove the exception and return null instead.
You could add again also the check
```
if (0 === \count($values)) {
    return null;
}
```

File: src/Admin/AbstractAdmin.php (hunk 2 above)
Comment (ggabrovski) [reply src/Admin/AbstractAdmin.php:2024]: > This shouldn't be called with `null`.
> 
> What's your use case ? There was none in the issue #6535

Because if you look in the template coming from [doctrine-orm-admin-bundle/Resources/views/Form/form_admin_fields.html.twig](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/5e602ba1871431ec3cc19e7421e674c5471b7e90/src/Resources/views/Form/form_admin_fields.html.twig#L59) on line 59 this function is called with `null` when you create new record. This can be fixed very easily if we make check in the template before calling the function.
Something like this:
`{% if sonata_admin.value and sonata_admin.field_description.associationadmin.id(sonata_admin.value) %}`
the same need to be done for line 62 where is the `objectId`.

I am quite new to Sonata and I don't know if you have other use cases like this one coming from other bundles and that's why I suggest to allow `null`. If you don`t have cases like this one we can fix the model list template. I can make PR to SonataDoctrineORMAdminBundle for this if you agree.

For the other things I will follow your advice and I will make new PRs.

Comment (VincentLanglet) [reply src/Admin/AbstractAdmin.php:2024]: IMHO 
```
{% if sonata_admin.value and sonata_admin.field_description.associationadmin.id(sonata_admin.value) %}
```
Is the right thing to do.

There is no reason to call an `id` method on a non-object. We shouldn't allow `null` just because it was (badly) used before. We're trying to improve our Interface, so it's better to fix the `id` calls.

I made https://github.com/sonata-project/SonataAdminBundle/pull/6578 on SonataAdmin to fix this on 3.x.
Can you do the PR on DoctrineORMAdmin ?
(The check is not needed line 62 because the code inside the if won't be executed with the check you'll add)

**Comment by ggabrovski** [comment]
I close this because of #6581 
Thanks for the support and guidelines.
